aws/v4/aws_v4.go

```diff
@@ -53,37 +54,29 @@ func (st Transport) RoundTrip(req *http.Request) (*http.Response, error) {
 		// Escaping path
 		req.URL.RawPath = url.PathEscape(req.URL.RawPath)
 	}
+
 	now := time.Now().UTC()
 	req.Header.Set("Date", now.Format(time.RFC3339))
+
 	var err error
 	switch req.Body {
 	case nil:
 		_, err = st.signer.Sign(req, nil, "es", st.region, now)
 	default:
-		buf, err := ioutil.ReadAll(req.Body)
-		if err != nil {
-			return nil, err
+		switch body := req.Body.(type) {
+		case io.ReadSeeker:
+			_, err = st.signer.Sign(req, body, "es", st.region, now)
+		default:
+			buf, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				return nil, err
+			}
+			req.Body = ioutil.NopCloser(bytes.NewReader(buf))
+			_, err = st.signer.Sign(req, bytes.NewReader(buf), "es", st.region, time.Now().UTC())
```

Comment:
To be honest, I don't think all of this is necessary at all. If I read [the documentation](https://github.com/aws/aws-sdk-go/blob/v1.30.7/aws/signer/v4/v4.go#L251-L275) correctly, the body should be automatically [wrapped in an `io.ReadCloser`](https://github.com/aws/aws-sdk-go/blob/v1.30.7/aws/signer/v4/v4.go#L261-L265):

```go
// Sign will set the request's Body to be the `body` parameter passed in. If
// the body is not already an io.ReadCloser, it will be wrapped within one. If
// a `nil` body parameter passed to Sign, the request's Body field will be
// also set to nil. Its important to note that this functionality will not
// change the request's ContentLength of the request.
```

Also, by default, the request body should be overwritten by the `Sign` method by default (as long as [`DisableRequestBodyOverwrite`](https://github.com/aws/aws-sdk-go/blob/v1.30.7/aws/signer/v4/v4.go#L190-L198) is `false`, which is the default):

```go
// Disables the automatical setting of the HTTP request's Body field with the
// io.ReadSeeker passed in to the signer. This is useful if you're using a
// custom wrapper around the body for the io.ReadSeeker and want to preserve
// the Body value on the Request.Body.
//
// This does run the risk of signing a request with a body that will not be
// sent in the request. Need to ensure that the underlying data of the Body
// values are the same.
DisableRequestBodyOverwrite bool
```

The code snippet that overrides the body is [here](https://github.com/aws/aws-sdk-go/blob/v1.30.7/aws/signer/v4/v4.go#L354-L366).

Do you have a test cluster on AWS where you could actually test this? I think it would help a lot in terms of removing allocation overhead if we could skip reading the request body.

Reply:
So, yes, I did just test this against an actual AWS ES cluster, and it does work.

I believe you are right, that we do not need to set the req.Body (line 74).
However, we do still need to read in the entire req.Body, and then wrap it in a bytes.Reader, because the AWS signer requires a io.ReadSeeker
The only way to get around that, is to have the main olivere/elastic library use ReadSeekers whenever possible (ie: bytes.Reader instead of bytes.Buffer).

Reply:
I can't assess the impact of changing `bytes.Reader` to `bytes.Buffer`, so I don't want to do this in the scope of this PR.

So, OK, if we can remove line 74 (wrapping with a nop closer) and the integration test with an actual AWS ES cluster worked, I'm happy to squash and merge. What do you think?

Reply:
I take that back, we can't remove line 74, because if DisableRequestBodyOverwrite is set to true, we will end up sending an empty request body (because it was already read).

So I think this PR is as good as it will get, without touching the main olivere/elastic files.

Reply:
Ok, well... I will merge when I'm back at my computer.